internal/nodes/avi_vrf_translator.go

```diff
@@ -31,42 +29,128 @@ import (
 	o.Lock.Lock()
 	defer o.Lock.Unlock()
-	aviVrfNode := &AviVrfNode{
-		Name: vrfName,
-	}
-
-	allNodes := objects.SharedNodeLister().CopyAllObjects()
-	// We need to sort the node list so that the staticroutes are in same order always
-	var nodeKeys []string
-	for k := range allNodes {
-		nodeKeys = append(nodeKeys, k)
+	//fetch vrf Node
+	aviVrfNodes := o.GetAviVRF()
+	if len(aviVrfNodes) == 0 {
+		vrfNode := &AviVrfNode{
+			Name: vrfName,
+		}
+		o.AddModelNode(vrfNode)
+		aviVrfNodes = append(aviVrfNodes, vrfNode)
 	}
-	sort.Strings(nodeKeys)
+	// Each AKO should have single VRF node as it deals with single cluster only.
+	aviVrfNode := aviVrfNodes[0]
 
-	utils.AviLog.Debugf("key: %s, All Nodes %v", key, allNodes)
 	routeid := 1
-	for _, k := range nodeKeys {
-		node := allNodes[k].(*v1.Node)
-		nodeRoutes, err := o.addRouteForNode(node, vrfName, routeid)
+	if len(aviVrfNode.StaticRoutes) != 0 {
+		routeid = len(aviVrfNode.StaticRoutes) + 1
+	} else {
+		aviVrfNode.NodeStaticRoutes = make(map[string]StaticRouteDetails)
+	}
+	var nodeRoutes []*models.StaticRoute
+	if !deleteFlag {
+		node, err := utils.GetInformers().NodeInformer.Lister().Get(nodeName)
+		if err != nil {
+			utils.AviLog.Errorf("key: %s, Error in fetching node details: %s: %v", key, nodeName, err)
+			return err
+		}
+		nodeRoutes, err = o.addRouteForNode(node, vrfName, routeid)
 		if err != nil {
-			utils.AviLog.Errorf("key: %s, Error Adding vrf for node %s: %v", key, node.Name, err)
-			continue
+			utils.AviLog.Errorf("key: %s, Error Adding vrf for node %s: %v", key, nodeName, err)
+			return err
 		}
+	}
+	// For new node addition (coming from ingestion layer), nodes static routes will be attahced at the end
+	// During reboot of AKO, nodes will be sorted. So this will give rest call
+	// HA case has to be checked: As Active and passive node, after failover, will not have same ordering of static
+	// routes
+	nodeStaticRouteDetails, ok := aviVrfNode.NodeStaticRoutes[nodeName]
+	if !ok {
+		//node not found, check overlapping and then add case
 		if !findRoutePrefix(nodeRoutes, aviVrfNode.StaticRoutes, key) {
+			// node is not present and no overlapping of cidr, append at last
 			aviVrfNode.StaticRoutes = append(aviVrfNode.StaticRoutes, nodeRoutes...)
-			routeid += len(nodeRoutes)
+			nodeStaticRoute := StaticRouteDetails{}
+			// start index shows at what index of StaticRoutes, nodes routes start (index based zero)
+			nodeStaticRoute.StartIndex = routeid - 1
+			nodeStaticRoute.Count = len(nodeRoutes)
+			aviVrfNode.NodeStaticRoutes[nodeName] = nodeStaticRoute
+			aviVrfNode.Nodes = append(aviVrfNode.Nodes, nodeName)
+		}
+	} else if !deleteFlag {
+		// update case
+		// Assumption: updated routes (values) for given node will not overlap with other nodes.
+		// So only updating existing routes of that node.
+		startIndex := nodeStaticRouteDetails.StartIndex
+		lenNewNodeRoutes := len(nodeRoutes)
+		diff := lenNewNodeRoutes - nodeStaticRouteDetails.Count
```

Comment:
shall we return early from here if diff is zero?

Reply:
There could be update to existing routes. So updating existing routes and then returning... instead of returning if we get difference zero.